### ISSUE TITLE
coerce str objects to the String datatype, not Unicode

### ISSUE THREAD

**Issue by zzzeek** [sqlalchemy/sqlalchemy#7551]
when we make an expression like this:

`literal('some string')`

the datatype should be String and not Unicode.  All DBAPIs accept Python string objects for VARCHAR etc. which are now all unicode.   The only difference that remains between String and Unicode is for big vendor DBs mssql and Oracle which imply NVARCHAR, NCLOB, etc.  These datatypes are known to be highly problematic when used in WHERE criteria etc. as they tend to cause indexes to be skipped and such.    

now that we are py3k only, the Unicode datatype should become a lesser used construct that's only required for the case when one wants to force "national" datatypes like NVARCHAR to be used, which is usually not the case.

**Comment by gordthompson** [comment]
So `literal('some string')` would default to `literal('some string', type_=String)` and `literal('some string', type_=Unicode)` would still be available if needed?

**Comment by zzzeek** [comment]
> So `literal('some string')` would default to `literal('some string', type_=String)` 

yes

> and `literal('some string', type_=Unicode)` would still be available if needed?

yes

SQL Server is the target here so that is better,right?

**Comment by gordthompson** [comment]
Yes. `WHERE nvarchar_col = 'foo'` can still use an index seek, whereas `WHERE varchar_col = N'foo'` will use an index scan.

**Comment by zzzeek** [comment]
oh interesting so it degrades in a good way in one direction and not the other, nifty

**Comment by zzzeek** [comment]
ah yes you are :)

try it like:

```
sa.bindparam(
                "foo_param",
                value="foo",
                type_=sa.String(255),
                literal_execute=True,
            )
```

though we're mostly looking at the behavior of real bound parameters, so

```
sa.bindparam(
                "foo_param",
                value="foo",
                type_=sa.String(255),
            )
```

**Comment by zzzeek** [comment]
but also, your example here:

```py
with engine.begin() as conn:
    results = conn.execute(
        sa.select(t.c.id).where(
            t.c.varchar_col == "foo"
        )
    ).fetchall()
```

we are not casting that as NVARCHAR.  that's just a Python string going into pyodbc.  if you look at the datatype, it's String.  so there's nothing that would change here implied by that:

```py
>>> from sqlalchemy import column, String
>>> expr = column('q', String()) == 'hi'
>>> expr.right
BindParameter('%(139660042673104 q)s', 'hi', type_=String())
```

the Unicode thing we are changing here only comes into play with there's no other comparison to be made:

```py
>>> from sqlalchemy import bindparam
>>> b = bindparam('x', value='y')
>>> b.type
Unicode()

>>> from sqlalchemy import literal
>>> l = literal('foo')
>>> l.type
Unicode()

```

if String is still being bound as N'' then that's that pyodbc setinputsizes() thing which I thought we had improved.

**Comment by zzzeek** [comment]
like try doing `create_engine("mssql+pyodbc://...", use_setinputsizes=True)`, see what that does now.

**Comment by gordthompson** [comment]
Okay, with

```python
engine = sa.create_engine(
    "mssql+pyodbc://scott:tiger^5HHH@localhost:49242/mydb"
    "?driver=ODBC+Driver+17+for+SQL+Server",
    use_setinputsizes=True,
    echo=True,
)
```

and stealing the logger example from the cx_Oracle dialect docs

```python
@event.listens_for(engine, "do_setinputsizes")
def _log_setinputsizes(inputsizes, cursor, statement, parameters, context):
    for bindparam, dbapitype in inputsizes.items():
            print(
                f"Bound parameter name: {bindparam.key}  "
                f"SQLAlchemy type: {bindparam.type}  "
                f"DBAPI object: {dbapitype}"
            )
```

then

```python
with engine.begin() as conn:
    results = conn.execute(
        sa.select(t.c.id).where(
            t.c.varchar_col == "foo"
        )
    ).fetchall()
```

shows

```
Bound parameter name: %(53007520576 varchar_col)s  SQLAlchemy type: VARCHAR(255) COLLATE "SQL_Latin1_General_CP1_CI_AS"  DBAPI object: <class 'str'>
```

and

```python
with engine.begin() as conn:
    results = conn.execute(
        sa.select(t.c.id).where(
            t.c.varchar_col == sa.literal("foo", type_=sa.String(255))
        )
    ).fetchall()
```

shows

```
Bound parameter name: %(637216790032 param)s  SQLAlchemy type: VARCHAR(255)  DBAPI object: <class 'str'>
```

What users will be looking for is the equivalent to this in pyodbc

```python
crsr = cnxn.cursor()
crsr.setinputsizes([(pyodbc.SQL_VARCHAR, 255)])
crsr.execute("SELECT * FROM million_rows WHERE varchar_col = ?", "foo")
```

because that winds up at the SQL Server as

```
declare @p1 int
set @p1=1
exec sp_prepexec @p1 output,N'@P1 varchar(255)',N'SELECT * FROM million_rows WHERE varchar_col = @P1','foo'
select @p1
```

**Comment by jvanasco** [comment]
> the Unicode datatype should become a lesser used construct

Would it make sense to emit a warning on PY3 when a `Unicode` column is instantiated or mapped?

My legacy projects migrated from PY2 all used `Unicode`. I might still use it in PY3 by habit.  Instructing people to adopt `String` now might be a good idea?

**Comment by gordthompson** [comment]
`Unicode` is still "a thing", at least for mssql and oracle. From what I've seen, `String` and `Unicode` and essentially synonyms for dialects where `varchar` is Unicode but, e.g., for mssql `String` → `varchar` and `Unicode` → `nvarchar`. Where it gets a bit messy is that in Py3 all `str` objects are Unicode.

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**establish mypy / typing approach for v2.0** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3562

**Comment by zzzeek** [comment]
OK so this breaks MSSQL tests immediately which are testing very exotic unicode characters.

so now I feel like I want to review what the original problem was here, where someone was complaining about indexes etc. because for a plain string comparison, "Column(String) == 'foo'" will use String for the right side.  It's only the literal() case, which is not common, that this matters.  So what are we doing here.

**Comment by zzzeek** [comment]
OK original complaints include:

#4561  didnt want the N, #4442 *does* want the N when Unicode is present, that's fixed, #4161 looks like it involves in py2 the pyodbc driver binding in a certain way and the N does not matter, #3111 im not sure what that is but it looks obsolete.

then gord you worked on https://stackoverflow.com/questions/55098426/strings-used-in-query-always-sent-with-nvarchar-syntax-even-if-the-underlying-c, this seems to be again pyodbc and not related to our side.

I think the bigger problem here is pyodbc assuming all Python strings should be N prefixed unless setinputsizes is used.   

one thing that we *could* do is get involved in the actual String/Unicode detection part of this.  Heh actually we can, we have a hook for this.  OK.  watch this

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**apply literal value resolution to String** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3577
